src/Storages/StorageMergeTree.cpp

```diff
@@ -457,6 +465,17 @@ Int64 StorageMergeTree::startMutation(const MutationCommands & commands, Context
         if (!inserted)
             throw Exception(ErrorCodes::LOGICAL_ERROR, "Mutation {} already exists, it's a bug", version);
 
+        if (partition)
+        {
+            last_mutation_by_partition[*partition] = version;
```

Comment:
Do we need to update this map on server startup as well? Please add an integration tests that runs mutation in partition and restarts server.

Reply:
Yes, it can improve performance in case there are lots of unfinished mutations for specific partitions, but strictly is not necessary. I'll make it in separate commit.

Reply:
bcfc588ed07e36e530b3f47e8d50e261cbd3f07b

Reply:
> Yes, it can improve performance in case there are lots of unfinished mutations for specific partitions, but strictly is not necessary

You're joking, right? According to the comment above you've implemented `partition_id` serialization in order to distinguish old and new IN PARTITION mutations after server restart. But it does not make any sense without actually using these `partition_id`s on startup.

Reply:
No, actually I implemented `partition_id` in order to have more straightforward approach and not calculate it every load (but only one time for mutation). But, it helped in distinguishing old and new mutations as well.

I'm not joking here, it actually not necessary for correctness, but improves performance in some use cases. This is so because if something is not found in map then search algorithm will simply go until the end of container (or `upper_bound()`, checking if `partition_id` is suitable. Basically, this map only saves iterations in case when there are many mutations for specific partitions, but does not affect correctness.